### PR TITLE
[Clang][Sema] Expose static inline functions from GMF

### LINKED ISSUE #98021: C++ modules appear to be exceedignly strict with intrinsic headers
That is, code using `immintrin.h` tends to fail to compile when using modules while working fine with headers.
I'll try to produce a minimal example in the next few hours.
For now, I have an example using boost_unordered.
When problems showed up in my own code using intrinsics, I could generally fix it by declaring all arguments as variables, and then passing the lvalues to the intriinsic function.

Hello.cxxm:
```c++
#ifndef USE_HEADERS
module;
#endif
#include <boost/unordered/unordered_flat_map.hpp>
#include <iostream>

#ifndef USE_HEADERS
export module Hello;
export {
#endif
  void hello() { std::cout << "Hello World!\n"; }
  template <typename K, typename V> using map = boost::unordered_flat_map<K, V>;
#ifndef USE_HEADERS
}
#endif
```
user.cpp:
```c++
#ifndef USE_HEADERS
import Hello;
#else
#include "hello.cxxm"
#endif

int main() {
  hello();
  int x = 0;
  long y = 0;
  map<int*,long*> m;
  m[&x] = &y;
  [[maybe_unused]] auto f = m.find(&x);
  return 0;
}
```

Compiling with headers:
```sh
$ clang++ -std=c++23 use.cpp -DUSE_HEADERS -o Hello.out
$ ./Hello.out
Hello World!
```
With modules:
```sh
$ clang++ -std=c++23 --precompile hello.cxxm -o M-hello.pcm
$ clang++ -std=c++23 use.cpp -fmodule-file=Hello=M-hello.pcm M-hello.pcm -o Hello_mod.out
```
results in
```
/usr/include/boost/unordered/detail/foa/core.hpp:293:7: error: no matching function for call to '_mm_cmpeq_epi8'
  293 |       _mm_cmpeq_epi8(load_metadata(),_mm_setzero_si128()))&0x7FFF;
      |       ^~~~~~~~~~~~~~
/usr/include/boost/unordered/detail/foa/core.hpp:309:14: note: in instantiation of member function 'boost::unordered::detail::foa::group15<boost::unordered::detail::foa::plain_integral>::match_available' requested here
  309 |     return (~match_available())&0x7FFF;
```
I could file this as a `boost_unordered` issue or make a PR there, as I've generally found I can work around the problem.
But I'll see about creating a minimal reproducer using `#include <immintrin.h>` directlry that works with headers but fails with modules.

### PR DESCRIPTION
**Pull request by jiixyj**
In C, it is a common pattern to have `static inline` functions in headers to avoid ODR issues. Currently, when those headers are included in a GMF, the names are not found when two-phase name lookup and ADL is involved. Those names are removed by `Sema::AddOverloadCandidate`.

Similarly, in C++, sometimes people use templates with internal linkage in headers.

As the GMF was designed to be a transitional mechanism for headers, special case those functions in `Sema::AddOverloadCandidate`.

This fixes <https://github.com/llvm/llvm-project/issues/98021>.

### CODE REVIEW

File: clang/lib/Sema/SemaOverload.cpp [hunk 1]
```diff
@@ -6926,11 +6926,26 @@ void Sema::AddOverloadCandidate(
...
+
+    /// Don't remove inline functions with internal linkage from the overload
+    /// set if they are declared in a GMF.
+    /// The global module is meant to be a transition mechanism for C and C++
+    /// headers.
+    /// Inline functions with internal linkage are a common pattern in headers
+    /// to avoid ODR issues.
+    const bool IsInlineFunctionInGMF =
+        Function->getOwningModule() &&
+        Function->getOwningModule()->isGlobalModule() &&
```
Comment (ChuanqiXu9) [inline clang/lib/Sema/SemaOverload.cpp:6945]:
```suggestion
        Function->isFromGlobalModule() 
```

File: clang/lib/Sema/SemaOverload.cpp [hunk 2]
```diff
@@ -6926,11 +6926,26 @@ void Sema::AddOverloadCandidate(
...
+      IsImplicitlyInstantiated = SpecInfo->getTemplateSpecializationKind() ==
+                                 TSK_ImplicitInstantiation;
+    }
+
+    /// Don't remove inline functions with internal linkage from the overload
+    /// set if they are declared in a GMF.
+    /// The global module is meant to be a transition mechanism for C and C++
+    /// headers.
+    /// Inline functions with internal linkage are a common pattern in headers
+    /// to avoid ODR issues.
```
Comment (ChuanqiXu9) [inline clang/lib/Sema/SemaOverload.cpp:6942]: I'd like to make the comments to be more explicitly to mention the behavior is not strictly conforming the standard. Maybe we need to add a NOTE. BTW, maybe this worth a paper to WG21. WDYT?

File: clang/lib/Sema/SemaOverload.cpp [hunk 3]
```diff
@@ -6926,11 +6926,26 @@ void Sema::AddOverloadCandidate(
...
+    const bool IsInlineFunctionInGMF =
+        Function->getOwningModule() &&
+        Function->getOwningModule()->isGlobalModule() &&
+        (IsImplicitlyInstantiated || Function->isInlined());
```
Comment (ChuanqiXu9) [inline clang/lib/Sema/SemaOverload.cpp:6946]: Maybe we can replace `(IsImplicitlyInstantiated || Function->isInlined())` with `getASTContext().GetGVALinkageForFunction(Function) == GVA_DiscardableODR`.

**Review by ChuanqiXu9 — APPROVED** [review]
LGTM except comments.

File: clang/lib/Sema/SemaOverload.cpp (hunk 3 above)
Comment (jiixyj) [reply clang/lib/Sema/SemaOverload.cpp:6945]: The GVA linkage seems to be sometimes `GVA_Internal` for some of the relevant functions, when running the new tests:

```c++
    if (IsImplicitlyInstantiated) {
      std::cerr << "IsImplicitlyInstantiated: GVA linkage: " << getASTContext().GetGVALinkageForFunction(Function) << '\n';
    }
    if (Function->isInlined()) {
      std::cerr << "Function->isInlined():    GVA linkage: " << getASTContext().GetGVALinkageForFunction(Function) << '\n';
    }
```

```
IsImplicitlyInstantiated: GVA linkage: 2
IsImplicitlyInstantiated: GVA linkage: 0
IsImplicitlyInstantiated: GVA linkage: 2
IsImplicitlyInstantiated: GVA linkage: 0
IsImplicitlyInstantiated: GVA linkage: 2
Function->isInlined():    GVA linkage: 0
IsImplicitlyInstantiated: GVA linkage: 2
Function->isInlined():    GVA linkage: 0
```

"0" is `GVA_Internal` and "2" is `GVA_DiscardableODR`. So for example `static inline` is `GVA_Internal`. So this wouldn't work I guess...

Comment (ChuanqiXu9) [reply clang/lib/Sema/SemaOverload.cpp:6945]: Yeah... this was the reason why it was rejected..

File: clang/lib/Sema/SemaOverload.cpp (hunk 2 above)
Comment (ChuanqiXu9) [reply clang/lib/Sema/SemaOverload.cpp:6942]: I feel the better wording will add a clause to https://eel.is/c++draft/basic.link#14. There are already some exceptions.

Comment (jiixyj) [reply clang/lib/Sema/SemaOverload.cpp]: Sorry for the delay! I have now updated the comment to include the fact that this fix/workaround violates [basic.link]p17, strictly speaking. I've also merged the main branch. On my machine, the (new) tests in `./tools/clang/test/Modules` still pass. Does this work for you? Thanks again for the review!

**Review by ChuanqiXu9 — APPROVED** [review]
LGTM. Thanks

**Comment by ChuanqiXu9** [comment]
You may want to add a note in ReleaseNotes too. And if you want to write a paper to WG21, please CC me or I'd love to be a co-author with you.

**Pull request merged**